src/genparticles.cxx

```diff
@@ -0,0 +1,305 @@
+            // 2. that the genparticle pt is larger than 8 GeV
+            // 3. the genparticle is isPrompt (statusbit 0) or
+            // isDirectPromptTauDecayProduct (statusbit 5)
+            bool statusbit = (IntBits(status_flags.at(i)).test(0) ||
+                              IntBits(status_flags.at(i)).test(5));
+            if ((pdgid == 11 || pdgid == 13) && pts.at(i) > 8 && statusbit) {
+                ROOT::Math::PtEtaPhiMVector gen_p4(pts.at(i), etas.at(i),
+                                                   phis.at(i), masses.at(i));
+                float delta_r =
+                    ROOT::Math::VectorUtil::DeltaR(gen_p4, lepton_p4);
+                if (delta_r < min_delta_r) {
+                    closest_genparticle_index = i;
+                    min_delta_r = delta_r;
+                }
+            }
+        }
+        Logger::get("genmatching::tau::genmatching")
+            ->debug("closted genelepton {} // DeltaR {}",
+                    closest_genparticle_index, min_delta_r);
+        // now loop trough the gentaus and check, if they are closer to the
+        // lepton than the closest lepton genparticle
+        for (auto hadronicGenTau : hadronicGenTaus) {
+            // check if the hadronicGenTau is closer to the lepton than the
+            // closest lepton genparticle
+            ROOT::Math::PtEtaPhiMVector hadronicGenTau_p4(
+                pts.at(hadronicGenTau), etas.at(hadronicGenTau),
+                phis.at(hadronicGenTau), masses.at(hadronicGenTau));
+            float gentau_delta_r =
+                ROOT::Math::VectorUtil::DeltaR(hadronicGenTau_p4, lepton_p4);
+            // the decay is considered a hadronic decay (statusbit 5) if
+            // 1. the hadronicGenTau pt is larger than 15 GeV
+            // 2. the delta_r is smaller than 0.2
+            // 3. the delta_r is smaller than the closest lepton genparticle
+            // delta_r
+            if (hadronicGenTau_p4.Pt() > 15 && gentau_delta_r < 0.2 &&
+                gentau_delta_r < min_delta_r) {
+                // statusbit 5 is hadronic tau decay
+                Logger::get("genmatching::tau::genmatching")
+                    ->debug(
+                        "found hadronicGenTau closer than closest lepton: {}",
+                        gentau_delta_r);
+                Logger::get("genmatching::tau::genmatching")
+                    ->debug("IS_TAU_HAD_DECAY");
+                return (int)GenMatchingCode::IS_TAU_HAD_DECAY;
+            }
+        }
+        // if it is not a hadronic decay, check if the lepton is close
+        // enough (deltaR < 0.2)
+        int closest_pdgid = std::abs(pdgids.at(closest_genparticle_index));
+        if (min_delta_r < 0.2) {
+            bool prompt =
+                IntBits(status_flags.at(closest_genparticle_index)).test(0);
+            bool from_tau =
+                IntBits(status_flags.at(closest_genparticle_index)).test(5);
```

Comment:
Is always only one true between prompt or from_tau, never both?

Reply:
it seems this way yes